documentation/specs/multithreading/analyzer-spec.md

```diff
@@ -0,0 +1,219 @@
+
+**Rationale**: No safe alternative. Tasks signal failure via return value/exception.
+
+### 2.2 MSB9998: TaskEnvironment Required
+
+**Severity**: Warning | **Detection**: Symbol matching | **Fixer**: Partial
+
+**APIs**:
+- `Environment.CurrentDirectory` → `TaskEnvironment.ProjectCurrentDirectory` (✅ fixer)
+- `Environment.GetEnvironmentVariable` → `TaskEnvironment.GetEnvironmentVariable` (✅ fixer)
+- `Environment.SetEnvironmentVariable(name, value)` → `TaskEnvironment.SetEnvironmentVariable` (✅ fixer)
+- `Path.GetFullPath` → `TaskEnvironment.GetAbsolutePath` (✅ fixer)
+- `Process.Start(...)`, `ProcessStartInfo` ctors → `TaskEnvironment.GetProcessStartInfo()` (❌ manual)
+
+### 2.3 MSB9997: File System APIs
+
+**Severity**: Warning | **Detection**: Pattern matching | **Fixer**: Yes
+
+**Types analyzed**: `File`, `Directory`, `FileInfo`, `DirectoryInfo`, `FileStream`, `StreamReader`, `StreamWriter`
+
+**Detection**: Warns if first `string` parameter is not:
+- Wrapped with `TaskEnvironment.GetAbsolutePath()`
+- A `.FullName` property
+- An `AbsolutePath` type
+
+**Example**:
+```csharp
+File.Exists(TaskEnvironment.GetAbsolutePath(path));  // ✅
+File.Exists(fileInfo.FullName);                       // ✅
+File.Exists(path);                                    // ❌ MSB9997
+```
+
+### 2.4 MSB9996: Potential Issues
+
+**Severity**: Warning | **Detection**: Symbol matching | **Fixer**: No
+
+**APIs**:
+- `Assembly.Load*`, `LoadFrom`, `LoadFile` - May cause conflicts
+- `Activator.CreateInstance*`, `AppDomain.Load/CreateInstance*` - May cause conflicts
+- `Console.*` (Write, WriteLine, ReadLine, etc.) - Interferes with logging
+
+**Rationale**: Not always wrong, requires case-by-case review.
+
+### 2.5 Limitations
+
+- No data-flow analysis (doesn't track if variable is absolute)
+- First-parameter heuristic only
+- No constant folding (warns on `"C:\\Windows"` literals)
+
+**Suppression**: `#pragma`, `.editorconfig`, or `[SuppressMessage]`
+
+---
+
+## 3. Code Fixers
+
+### 3.1 MSB9998: API Replacements
+
+```csharp
+Environment.CurrentDirectory       → TaskEnvironment.ProjectCurrentDirectory
+Environment.GetEnvironmentVariable → TaskEnvironment.GetEnvironmentVariable  
+Environment.SetEnvironmentVariable → TaskEnvironment.SetEnvironmentVariable
+Path.GetFullPath                   → TaskEnvironment.GetAbsolutePath
+```
+
+### 3.2 MSB9997: Path Wrapping
+
+Wraps first `string` argument:
+
+```csharp
+- File.Exists(somePath)
++ File.Exists(TaskEnvironment.GetAbsolutePath(somePath))
+```
+
+### 3.3 User Experience
+
+**IDE**: Quick Actions (Ctrl+.) with "Fix All" support  
+**CLI**: `dotnet format analyzers --diagnostics MSB9997 MSB9998`
+
+---
+
+## 4. Open Questions
+
+### 4.1 Distribution
+
+**Proposal**: Ship with `Microsoft.Build.Utilities.Core` NuGet package
+
+**Pros**: Already referenced, zero config, automatic updates  
+**Cons**: Increases package size
+
+**Question**: Consensus on Utilities.Core vs standalone package?
```

Comment:
I think I want a new package like `Microsoft.Build.TaskAuthoring` that

* has logic to make sure you're not unnecessarily redistributing MSBuild assemblies
* Has this analyzer
* provides the API-surface references as needed
* maybe in the future some test thing?